modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/actions/PreIssueAccessTokenCodeGrantTestCase.java

```diff
@@ -0,0 +1,701 @@
+        RoleV2 role = new RoleV2(roleAudience, TEST_ROLE_APPLICATION, userPermissions, Collections.emptyList());
+        roleId = addRole(role);
+
+        // Creates user
+        UserObject userInfo = new UserObject();
+        userInfo.setUserName(TEST_USER);
+        userInfo.setPassword(ADMIN_WSO2);
+        userInfo.setName(new Name().givenName(TEST_USER_FIRST_NAME));
+        userInfo.getName().setFamilyName(TEST_USER_LAST_NAME);
+        userInfo.addEmail(new Email().value(TEST_USER_GMAIL_COM));
+        userId = scim2RestClient.createUser(userInfo);
+
+        // Assigns role to the created user
+        RoleItemAddGroupobj rolePatchReqObject = new RoleItemAddGroupobj();
+        rolePatchReqObject.setOp(RoleItemAddGroupobj.OpEnum.ADD);
+        rolePatchReqObject.setPath(USERS);
+        rolePatchReqObject.addValue(new ListObject().value(userId));
+        scim2RestClient.updateUserRole(new PatchOperationRequestObject().addOperations(rolePatchReqObject), roleId);
+    }
+
+    /**
+     * Adds permissions based on the provided custom scopes.
+     *
+     * @param customScopes A list of custom scopes to add as permissions
+     * @return A list of permissions including both predefined and custom scope-based permissions
+     */
+    private List<Permission> addPermissions(List<String> customScopes) {
+
+        List<Permission> userPermissions = new ArrayList<>();
+
+        Collections.addAll(userPermissions,
+                new Permission(INTERNAL_ORG_USER_MANAGEMENT_LIST),
+                new Permission(INTERNAL_ORG_USER_MANAGEMENT_VIEW),
+                new Permission(INTERNAL_ORG_USER_MANAGEMENT_CREATE),
+                new Permission(INTERNAL_ORG_USER_MANAGEMENT_UPDATE),
+                new Permission(INTERNAL_ORG_USER_MANAGEMENT_DELETE)
+                          );
+
+        customScopes.forEach(scope -> userPermissions.add(new Permission(scope)));
+
+        return userPermissions;
+    }
+
+    private String createOIDCAppWithClaims() throws Exception {
+
+        List<UserClaimConfig> userClaimConfigs = Arrays.asList(
+                new UserClaimConfig.Builder().localClaimUri("http://wso2.org/claims/givenname").
+                        oidcClaimUri(FIRST_NAME_CLAIM).build(),
+                new UserClaimConfig.Builder().localClaimUri("http://wso2.org/claims/lastname").
+                        oidcClaimUri(LAST_NAME_CLAIM).build()
+                                                              );
+
+        ApplicationConfig applicationConfig = new ApplicationConfig.Builder()
+                .claimsList(userClaimConfigs)
+                .grantTypes(new ArrayList<>(Collections.singleton(OAUTH2_GRANT_TYPE_AUTHORIZATION_CODE)))
+                .tokenType(ApplicationConfig.TokenType.JWT)
+                .expiryTime(3600)
+                .skipConsent(true)
+                .build();
+
+        appTokenScopes = new TokenScopes.Builder().requestedScopes(Arrays.asList(OPENID_SCOPE, PROFILE_SCOPE,
```

Comment:
Why do we have this inside this method ? And can't see any usages of this.

Reply:
This is refactored with e641385fd8a50b37b7b3090ff41135d479266800